File: tests/wasm/test_fibonacci_entry_bands.js
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { assertArrayClose, isNaN } from './test_utils.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;

function sampleOhlc(length = 256) {
    const open = new Float64Array(length);
    const high = new Float64Array(length);
    const low = new Float64Array(length);
    const close = new Float64Array(length);
    for (let i = 0; i < length; i++) {
        open[i] = 100 + i * 0.05 + Math.sin(i * 0.09) * 1.3;
        close[i] = open[i] + Math.cos(i * 0.07) * 1.1;
        high[i] = Math.max(open[i], close[i]) + 0.65 + Math.abs(Math.sin(i * 0.03)) * 0.25;
        low[i] = Math.min(open[i], close[i]) - 0.65 - Math.abs(Math.cos(i * 0.05)) * 0.25;
    }
    return { open, high, low, close };
}

function outputKeys() {
    return [
        'basis',
        'trend',
        'upper_0618',
        'upper_1000',
        'upper_1618',
        'upper_2618',
        'lower_0618',
        'lower_1000',
        'lower_1618',
        'lower_2618',
        'tp_long_band',
        'tp_short_band',
        'long_entry',
        'short_entry',
        'rejection_long',
        'rejection_short',
        'long_bounce',
        'short_bounce',
    ];
}

test.before(async () => {
    const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
    const importPath =
        process.platform === 'win32'
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
    const imported = await import(importPath);
    wasm = imported.default ?? imported;
});

test('fibonacci_entry_bands_js output contract', () => {
    const { open, high, low, close } = sampleOhlc();
    const out = wasm.fibonacci_entry_bands_js(open, high, low, close, 'hlc3', 21, 14, true, 'low');

    for (const key of outputKeys()) {
        assert.strictEqual(out[key].length, close.length);
    }

    assert(out.basis.some(v => !isNaN(v)));
    assert(out.tp_long_band.some(v => !isNaN(v)));
    assert(out.tp_short_band.some(v => !isNaN(v)));

    for (const value of out.trend.filter(v => !isNaN(v))) {
        assert(value === -1 || value === 0 || value === 1);
    }

    for (const key of [
        'long_entry',
        'short_entry',
        'rejection_long',
        'rejection_short',
        'long_bounce',
        'short_bounce',
    ]) {
        for (const value of out[key].filter(v => !isNaN(v))) {
            assert(value === 0 || value === 1);
        }
    }
});

test('fibonacci_entry_bands_js rejects invalid inputs', () => {
    const { open, high, low, close } = sampleOhlc(32);
    assert.throws(
        () => wasm.fibonacci_entry_bands_js(open.subarray(0, 16), high, low, close, 'hlc3', 21, 14, true, 'low'),
        /Inconsistent slice lengths/,
    );
    assert.throws(
        () => wasm.fibonacci_entry_bands_js(open, high, low, close, 'bad', 21, 14, true, 'low'),
        /Invalid source/,
    );
    assert.throws(
        () => wasm.fibonacci_entry_bands_js(open, high, low, close, 'hlc3', 0, 14, true, 'low'),
        /Invalid length/,
    );
    assert.throws(
        () => wasm.fibonacci_entry_bands_js(open, high, low, close, 'hlc3', 21, 14, true, 'bad'),
        /Invalid TP aggressiveness/,
    );
});

test('fibonacci_entry_bands_into pointer path matches safe API', () => {
    const { open, high, low, close } = sampleOhlc(128);
    const safe = wasm.fibonacci_entry_bands_js(open, high, low, close, 'hlc3', 16, 11, true, 'high');

    const openPtr = wasm.copy_f64_array(open);
    const highPtr = wasm.copy_f64_array(high);
    const lowPtr = wasm.copy_f64_array(low);
    const closePtr = wasm.copy_f64_array(close);
    const ptrs = {};
    for (const key of outputKeys()) {
        ptrs[key] = wasm.fibonacci_entry_bands_alloc(close.length);
    }

    try {
        wasm.fibonacci_entry_bands_into(
            openPtr,
            highPtr,
            lowPtr,
            closePtr,
            ptrs.basis,
            ptrs.trend,
            ptrs.upper_0618,
            ptrs.upper_1000,
            ptrs.upper_1618,
            ptrs.upper_2618,
            ptrs.lower_0618,
            ptrs.lower_1000,
            ptrs.lower_1618,
            ptrs.lower_2618,
            ptrs.tp_long_band,
            ptrs.tp_short_band,
            ptrs.long_entry,
            ptrs.short_entry,
            ptrs.rejection_long,
            ptrs.rejection_short,
            ptrs.long_bounce,
            ptrs.short_bounce,
            close.length,
            'hlc3',
            16,
            11,
            true,
            'high',
        );

        for (const key of outputKeys()) {
            assertArrayClose(
                wasm.read_f64_array(ptrs[key], close.length),
                safe[key],
                1e-12,
                key + ' mismatch',
            );
        }
    } finally {
        for (const key of outputKeys()) {
            wasm.fibonacci_entry_bands_free(ptrs[key], close.length);
        }
        wasm.deallocate_f64_array(openPtr);
        wasm.deallocate_f64_array(highPtr);
        wasm.deallocate_f64_array(lowPtr);
        wasm.deallocate_f64_array(closePtr);
    }
});

test('fibonacci_entry_bands_js resets after NaN', () => {
    const { open, high, low, close } = sampleOhlc(220);
    open[90] = Number.NaN;
    high[90] = Number.NaN;
    low[90] = Number.NaN;
    close[90] = Number.NaN;

    const out = wasm.fibonacci_entry_bands_js(open, high, low, close, 'hlc3', 16, 11, true, 'high');
    for (const key of outputKeys()) {
        assert(isNaN(out[key][90]), key + ' expected NaN at reset point');
    }
});

test('fibonacci_entry_bands_batch_js single parameter set matches safe API', () => {
    const { open, high, low, close } = sampleOhlc(160);
    const batch = wasm.fibonacci_entry_bands_batch_js(open, high, low, close, {
        length_range: [17, 17, 0],
        atr_length_range: [10, 10, 0],
        source: 'hlc3',
        use_atr: true,
        tp_aggressiveness: 'medium',
    });
    const single = wasm.fibonacci_entry_bands_js(open, high, low, close, 'hlc3', 17, 10, true, 'medium');

    assert.strictEqual(batch.rows, 1);
    assert.strictEqual(batch.cols, close.length);
    assert.deepStrictEqual(batch.lengths, [17]);
    assert.deepStrictEqual(batch.atr_lengths, [10]);
    assert.deepStrictEqual(batch.sources, ['hlc3']);
    assert.deepStrictEqual(batch.use_atr_flags, [true]);
    assert.deepStrictEqual(batch.tp_aggressiveness_values, ['medium']);

    for (const key of outputKeys()) {
        assertArrayClose(batch[key], single[key], 1e-12, key + ' batch mismatch');
    }
});

test('fibonacci_entry_bands_batch_js metadata and batch_into', () => {
    const { open, high, low, close } = sampleOhlc(120);
    const batch = wasm.fibonacci_entry_bands_batch_js(open, high, low, close, {
        length_range: [16, 18, 2],
        atr_length_range: [10, 12, 2],
        source: 'hl2',
        use_atr: false,
        tp_aggressiveness: 'low',
    });

    assert.strictEqual(batch.rows, 4);
    assert.strictEqual(batch.cols, close.length);
    assert.deepStrictEqual(batch.lengths, [16, 16, 18, 18]);
    assert.deepStrictEqual(batch.atr_lengths, [10, 12, 10, 12]);
    assert.deepStrictEqual(batch.sources, ['hl2', 'hl2', 'hl2', 'hl2']);
    assert.deepStrictEqual(batch.use_atr_flags, [false, false, false, false]);
    assert.deepStrictEqual(batch.tp_aggressiveness_values, ['low', 'low', 'low', 'low']);

    const openPtr = wasm.copy_f64_array(open);
    const highPtr = wasm.copy_f64_array(high);
    const lowPtr = wasm.copy_f64_array(low);
    const closePtr = wasm.copy_f64_array(close);
    const total = batch.rows * close.length;
    const ptrs = {};
    for (const key of outputKeys()) {
        ptrs[key] = wasm.fibonacci_entry_bands_alloc(total);
    }

    try {
        const rows = wasm.fibonacci_entry_bands_batch_into(
            openPtr,
            highPtr,
            lowPtr,
            closePtr,
            ptrs.basis,
            ptrs.trend,
            ptrs.upper_0618,
            ptrs.upper_1000,
            ptrs.upper_1618,
            ptrs.upper_2618,
            ptrs.lower_0618,
            ptrs.lower_1000,
            ptrs.lower_1618,
            ptrs.lower_2618,
            ptrs.tp_long_band,
            ptrs.tp_short_band,
            ptrs.long_entry,
            ptrs.short_entry,
            ptrs.rejection_long,
            ptrs.rejection_short,
            ptrs.long_bounce,
            ptrs.short_bounce,
            close.length,
            16,
            18,
            2,
            10,
            12,
            2,
            'hl2',
            false,
            'low',
        );
        assert.strictEqual(rows, 4);

        for (const key of outputKeys()) {
            assertArrayClose(
                wasm.read_f64_array(ptrs[key], total),
                batch[key],
                1e-12,
                key + ' batch_into mismatch',
            );
        }
    } finally {
        for (const key of outputKeys()) {
            wasm.fibonacci_entry_bands_free(ptrs[key], total);
        }
        wasm.deallocate_f64_array(openPtr);
        wasm.deallocate_f64_array(highPtr);
        wasm.deallocate_f64_array(lowPtr);
        wasm.deallocate_f64_array(closePtr);
    }
});

test('fibonacci_entry_bands_batch_js rejects invalid config', () => {
    const { open, high, low, close } = sampleOhlc(32);
    assert.throws(
        () => wasm.fibonacci_entry_bands_batch_js(open, high, low, close, { length_range: 'bad' }),
        /Invalid config/,
    );
});

File: tests/python/test_fibonacci_entry_bands.py
import numpy as np
import pytest

try:
    import vector_ta as mp
except ImportError:
    try:
        import vector_ta as mp
    except ImportError:
        pytest.skip(
            "Python module not built. Run 'maturin develop --features python' first",
            allow_module_level=True,
        )


def sample_ohlc(length=256):
    idx = np.arange(length, dtype=np.float64)
    open_ = 100.0 + idx * 0.05 + np.sin(idx * 0.09) * 1.3
    close = open_ + np.cos(idx * 0.07) * 1.1
    high = np.maximum(open_, close) + 0.65 + np.abs(np.sin(idx * 0.03)) * 0.25
    low = np.minimum(open_, close) - 0.65 - np.abs(np.cos(idx * 0.05)) * 0.25
    return open_, high, low, close


def output_keys():
    return [
        "basis",
        "trend",
        "upper_0618",
        "upper_1000",
        "upper_1618",
        "upper_2618",
        "lower_0618",
        "lower_1000",
        "lower_1618",
        "lower_2618",
        "tp_long_band",
        "tp_short_band",
        "long_entry",
        "short_entry",
        "rejection_long",
        "rejection_short",
        "long_bounce",
        "short_bounce",
    ]


class TestFibonacciEntryBands:
    def test_output_contract(self):
        open_, high, low, close = sample_ohlc()
        outputs = mp.fibonacci_entry_bands(
            open_, high, low, close, source="hlc3", length=21, atr_length=14, use_atr=True
        )

        assert set(outputs.keys()) == set(output_keys())
        for key in output_keys():
            assert len(outputs[key]) == len(close)

        assert np.isfinite(outputs["basis"]).any()
        assert np.isfinite(outputs["tp_long_band"]).any()
        assert np.isfinite(outputs["tp_short_band"]).any()

        trend = outputs["trend"]
        trend_valid = trend[np.isfinite(trend)]
        assert np.all((trend_valid == -1.0) | (trend_valid == 0.0) | (trend_valid == 1.0))

        for key in [
            "long_entry",
            "short_entry",
            "rejection_long",
            "rejection_short",
            "long_bounce",
            "short_bounce",
        ]:
            signal = outputs[key]
            valid = signal[np.isfinite(signal)]
            assert np.all((valid == 0.0) | (valid == 1.0))

    def test_invalid_parameters(self):
        open_, high, low, close = sample_ohlc(32)

        with pytest.raises(ValueError, match="Invalid source"):
            mp.fibonacci_entry_bands(open_, high, low, close, source="bad")

        with pytest.raises(ValueError, match="Invalid length"):
            mp.fibonacci_entry_bands(open_, high, low, close, length=0)

        with pytest.raises(ValueError, match="Invalid TP aggressiveness"):
            mp.fibonacci_entry_bands(open_, high, low, close, tp_aggressiveness="bad")

    def test_stream_matches_batch_with_reset(self):
        open_, high, low, close = sample_ohlc(220)
        open_[90] = np.nan
        high[90] = np.nan
        low[90] = np.nan
        close[90] = np.nan

        batch = mp.fibonacci_entry_bands(
            open_,
            high,
            low,
            close,
            source="hlc3",
            length=16,
            atr_length=11,
            use_atr=True,
            tp_aggressiveness="high",
        )
        stream = mp.FibonacciEntryBandsStream(
            source="hlc3",
            length=16,
            atr_length=11,
            use_atr=True,
            tp_aggressiveness="high",
        )

        streamed = {key: [] for key in output_keys()}
        for values in zip(open_, high, low, close):
            out = stream.update(*map(float, values))
            if out is None:
                for key in output_keys():
                    streamed[key].append(np.nan)
            else:
                for key in output_keys():
                    streamed[key].append(out[key])

        assert stream.warmup_period == 10
        for key in output_keys():
            np.testing.assert_allclose(
                batch[key],
                streamed[key],
                rtol=1e-12,
                atol=1e-12,
                equal_nan=True,
            )

    def test_batch_single_row_matches_single(self):
        open_, high, low, close = sample_ohlc(128)
        batch = mp.fibonacci_entry_bands_batch(
            open_,
            high,
            low,
            close,
            length_range=(17, 17, 0),
            atr_length_range=(10, 10, 0),
            source="hlc3",
            use_atr=True,
            tp_aggressiveness="medium",
        )
        single = mp.fibonacci_entry_bands(
            open_,
            high,
            low,
            close,
            source="hlc3",
            length=17,
            atr_length=10,
            use_atr=True,
            tp_aggressiveness="medium",
        )

        assert batch["rows"] == 1
        assert batch["cols"] == len(close)
        np.testing.assert_array_equal(batch["lengths"], np.array([17], dtype=np.uint64))
        np.testing.assert_array_equal(batch["atr_lengths"], np.array([10], dtype=np.uint64))
        np.testing.assert_array_equal(batch["use_atr_flags"], np.array([True]))
        assert batch["sources"] == ["hlc3"]
        assert batch["tp_aggressiveness_values"] == ["medium"]

        for key in output_keys():
            np.testing.assert_allclose(
                batch[key][0], single[key], rtol=1e-12, atol=1e-12, equal_nan=True
            )

    def test_batch_metadata(self):
        open_, high, low, close = sample_ohlc(120)
        batch = mp.fibonacci_entry_bands_batch(
            open_,
            high,
            low,
            close,
            length_range=(16, 18, 2),
            atr_length_range=(10, 12, 2),
            source="hl2",
            use_atr=False,
            tp_aggressiveness="low",
        )

        assert batch["rows"] == 4
        assert batch["cols"] == len(close)
        for key in output_keys():
            assert batch[key].shape == (4, len(close))

        np.testing.assert_array_equal(
            batch["lengths"], np.array([16, 16, 18, 18], dtype=np.uint64)
        )
        np.testing.assert_array_equal(
            batch["atr_lengths"], np.array([10, 12, 10, 12], dtype=np.uint64)
        )
        np.testing.assert_array_equal(batch["use_atr_flags"], np.array([False, False, False, False]))
        assert batch["sources"] == ["hl2", "hl2", "hl2", "hl2"]
        assert batch["tp_aggressiveness_values"] == ["low", "low", "low", "low"]


if __name__ == "__main__":
    pytest.main([__file__, "-v"])
